### PR TITLE
GCI: Ensure that the right version of kubelet is used

### PR DESCRIPTION
**Pull request by andyzheng0831**
This PR fixes a binary version issue in using GCI image. Unlike ContainerVM, we don't hard-code a GCI image version in kubernetes code. When a user creates a cluster using GCI, by default the latest "gci-dev" will be used, in which the latest kubelet and kubectl release binary are built. The problem is that the user may not intend to use the latest version. This PR is for fixing it.

@roberthbailey @dchen1107 please review it or assign it to a suitable reviewer. Please also mark it as a cherry-pick candidate.

cc/ @fabioy @wonderfly @kubernetes/goog-image

### CODE REVIEW

File: cluster/gce/trusty/configure.sh [hunk 1]
```diff
@@ -122,17 +122,34 @@ install_kube_binary_config() {
...
-    mount --bind -o remount,ro,^noexec "${BIN_PATH}/kubectl" "${BIN_PATH}/kubectl"
+    chmod 544 "${kube_bin}/kubelet"
+    chmod 544 "${kube_bin}/kubectl"
+    # If the built-in binary version is different from the expected version, we use
+    # the downloaded binary. The simplest implementation is to always use the downloaded
+    # binary without checking the version. But we have another version guardian in GKE.
+    # So, we compare the versions to ensure this run-time binary replacement is only
+    # applied for OSS kubernetes.
+    readonly builtin_version="$(/usr/bin/kubelet --version=true | cut -f2 -d " ")"
+    readonly required_version="$(/home/kubernetes/bin/kubelet --version=true | cut -f2 -d " ")"
```
Comment (adityakali) [inline cluster/gce/trusty/configure.sh]: shouldn't this be taken from some environment var?

File: cluster/gce/trusty/configure.sh [hunk 2]
```diff
@@ -122,17 +122,34 @@ install_kube_binary_config() {
...
+    if [ "${TEST_CLUSTER:-}" = "true" ] || [ "${builtin_version}" != "${required_version}" ]; then
+      mount --bind "${kube_bin}/kubelet" "${BIN_PATH}/kubelet"
+      mount --bind "${kube_bin}/kubectl" "${BIN_PATH}/kubectl"
+    else
+      # Remove downloaded binary just to prevent misuse.
```
Comment (adityakali) [inline cluster/gce/trusty/configure.sh]: Don't delete these. If the built-in version changes coz of auto-update, then you will need to use these binaries. I guess you will probably end up downloading everything again, but there is no harm in leaving them for possible optimization in future.

Comment (andyzheng0831) [inline cluster/gce/trusty/configure.sh]: I don't think so. Clean up unnecessary kubelet and kubectl makes thing less confuse. When people see /home/kubernetes/bin/kubelet, they will not expect it is different from the binary in use. As for optimization, it is not straightforward at here. We need to make a plan and then do it in dedicated PRs.

File: cluster/gce/trusty/configure.sh (hunk 1 above)
Comment (andyzheng0831) [inline cluster/gce/trusty/configure.sh]: Originally, I thought we can get version info from kube-env. But later I noticed there is no version var in that file. Theoretically, we can add KUBE_VERSION in kube-env file. There two downsides: (1) it will change much more places; (2) the meaning of KUBE_VERSION itself is complex (https://github.com/kubernetes/kubernetes/blob/master/cluster/common.sh#l257). A similar way is get version from the tarball URL listed in kube-env, but it has the same problem. So, the simplest way is to run "kubelet --version" just like what ContainerVM's config does (https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/configure-vm.sh#l89)

**Comment by andyzheng0831** [comment]
@roberthbailey will you please take a look? We will need to cherry pick it to release-1.2 before the next 1.2.X release. Thanks

File: cluster/gce/trusty/configure.sh (hunk 1 above)
Comment (roberthbailey) [inline cluster/gce/trusty/configure.sh]: /cc @zmerlynn

**Comment by roberthbailey** [comment]
This looks ok to me but I'm going to assign to @zmerlynn to take a look since I won't be able to review it in the next couple of days.

**Comment by lavalamp** [comment]
Do we have an e2e job that tests this? (build cop wondering how safe this is to merge)

**Comment by andyzheng0831** [comment]
This change only affects GCI. We don't have a e2e job to test a pending PR. But I had manually ran e2e to verify it. And also ssh into instances to verify the logic. It is safe to merge I think

**Comment by lavalamp** [comment]
@andyzheng0831 In the future, if you know something won't be tested by our e2e tests, please add the "e2e-not-required" label, otherwise the submit queue runs a pointless 40 minute test on it.

**Comment by andyzheng0831** [comment]
@lavalamp Thanks for the suggestion! That will be better for both submit queue and GCI-specific changes
